tools/azure-pipelines/jobs-template.yml

```diff
@@ -0,0 +1,137 @@
+
+  steps:
+  # The cache task is persisting the .m2 directory between builds, so that
+  # we do not have to re-download all dependencies from maven central for 
+  # each build. The hope is that downloading the cache is faster than
+  # all dependencies individually.
+  # In this configuration, we a hash over all committed (not generated) .pom files 
+  # as a key for the build cache (CACHE_KEY). If we have a cache miss on the hash
+  # (usually because a pom file has changed), we'll fall back to a key without
+  # the pom files (CACHE_FALLBACK_KEY).
+  # Offical documentation of the Cache task: https://docs.microsoft.com/en-us/azure/devops/pipelines/caching/?view=azure-devops
+  - task: Cache@2
+    inputs:
+      key: $(CACHE_KEY)
+      restoreKeys: $(CACHE_FALLBACK_KEY)
+      path: $(MAVEN_CACHE_FOLDER)
+    continueOnError: true # continue the build even if the cache fails.
+    displayName: Cache Maven local repo
+
+  # Compile
+  - script: STAGE=compile ${{parameters.environment}} ./tools/azure_controller.sh compile
+    displayName: Build
+
+  # upload artifacts for next stage
+  - task: PublishPipelineArtifact@1
+    inputs:
+      path: $(CACHE_FLINK_DIR)
+      artifact: FlinkCompileCacheDir-${{parameters.stage_name}}
+
+- job: test_${{parameters.stage_name}}
+  dependsOn: compile_${{parameters.stage_name}}
+  condition: not(eq(variables['MODE'], 'e2e'))
+  pool: ${{parameters.test_pool_definition}}
+  container: flink-build-container
+  timeoutInMinutes: 240
+  cancelTimeoutInMinutes: 1
+  workspace:
+    clean: all
+  strategy:
+    matrix:
+      core:
+        module: core
+      python:
+        module: python
+      libraries:
+        module: libraries
+      blink_planner:
+        module: blink_planner
+      connectors:
+        module: connectors
+      kafka_gelly:
+        module: kafka/gelly
+      tests:
+        module: tests
+      legacy_scheduler_core:
+        module: legacy_scheduler_core
+      legacy_scheduler_tests:
+        module: legacy_scheduler_tests
+      misc:
+        module: misc
+  steps:
+
+  # download artifacts
+  - task: DownloadPipelineArtifact@2
+    inputs:
+      path: $(CACHE_FLINK_DIR)
+      artifact: FlinkCompileCacheDir-${{parameters.stage_name}}
+
+  # recreate "build-target" symlink for python tests
+  - script: |
+      ls -lisah $(CACHE_FLINK_DIR)
+      ls -lisah .
+      ln -snf $(CACHE_FLINK_DIR)/flink-dist/target/flink-*-SNAPSHOT-bin/flink-*-SNAPSHOT $(CACHE_FLINK_DIR)/build-target
+    displayName: Recreate 'build-target' symlink
+  # Test
+  - script: STAGE=test ${{parameters.environment}} ./tools/azure_controller.sh $(module)
+    displayName: Test - $(module)
+
+  - task: PublishTestResults@2
+    inputs:
+      testResultsFormat: 'JUnit'
+
+
+- job: e2e_${{parameters.stage_name}}
+  condition: eq(variables['MODE'], 'e2e')
+  # We are not running this job on a container, but in a VM.
```

Comment:
how come? Isn't there a reference somewhere else that some e2e tests use docker and hence the container must be setup outside?

Reply:
the end to end tests are fairly invasive on the underlying system, that's why I wanted to use ephemeral machines. Also since the e2e tests use docker extensively, we would have much more docker in docker trouble

Reply:
But the pre-commit-tests are still being run in docker, no?

azure-controller.sh#L191: `Some tests in the "run-pre-commit-tests.sh" collection launch Docker containers.`

If running end-to-end tests in a docker container is troublesome, shouldn't the pre-commit-tests also be executed separately? As it stands we have to handle both setups.

Reply:
Correct, the pre-commit tests run in docker (because they are executed on each push / pr). For now, I want all per push / pr stuff running on the custom build machines.
If we are going to always executed the end to end tests, we can consider moving the pre-commit tests.

Reply:
Shouldn't we consider doing that now since it would make several changes in this PR unnecessary?

Reply:
I'm trying out running the precommit tests in a VM: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=5144&view=results

Reply:
Running the pre commit on a separate machine doesn't make much sense, as it needs 25 minutes of compiling, and then runs for 2 minutes: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=5164&view=logs&j=62470212-5e37-5318-626e-1835b0c31696

I'm now trying to run the entire misc stage on an azure hosted instance.

Reply:
why isn't it re-using the cached compilation?

Reply:
And these numbers still look _fine_ to me; they aren't increasing the total build duration.

Reply:
FYI; you may want to look into decreasing the transfer.sh timeout, we blew 10 minutes in multiple profiles on that build you linked.

Reply:
Didn't the build take so long because it had to download all dependencies?

Reply:
I kicked off this hacky build last night, that's why some things are not correct.
I now include the custom maven version setup step, and the cache: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=5168&view=results
Let's see how it affects the build time.

Reply:
It was again downloading all dependencies.
It's not compiling Flink (so that caching works), but the maven caching seems broken.

Reply:
This build is also still not using maven 3.2.5

Reply:
I will look into the problems you've mentioned. This is a successful run of running the entire misc stage on an Azure instance: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=5172&view=results
The benefit of this approach would be, that the number of builds stays the same.

Reply:
yeah but the build times for misc has increased by 17 minutes, increasing the total build duration if all jobs are running.
I think in your previous version you also weren't running the java pre-commit-tests, which make a dedicated job more useful.

Reply:
I fixed the downloading of the dependencies, and it uses maven 3.2.5: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=5189&view=results